### ISSUE TITLE
OrientDB v3.0.x - Retrieving vertex from indexed property in local transaction is slow

### ISSUE THREAD

**Issue by CyprienGottstein** [orientechnologies/orientdb#8329]
### OS: Ubuntu 16.04 

Hello OrientDB Team !

I spotted a new issue concerning heavy bulk insertion and using index on local transaction.

In our tests suite we want to bulk insert a heavy payload of 10k vertices, we could do things in an other way but since it worked in v2.2.30, I'm raising this issue. To avoid data duplication, we perform checks on an indexed property with a unique index.

The checks were performed in constant time in v2.2.30 with an average time of 0/1 ms, now the time needed to do a single vertex checks is growing in a linear way and can go from 0/1 ms to more than 50ms which is quite problematic on our side.

I've created a gist with the appropriate snippet which shows the problem, spatial data are not just for show we have this use case in our project.

https://gist.github.com/CyprienGottstein/53b8c66578d453cec236c75d84f44ac3

The growing time of the index check is, in my guess, linked to the size of the current transaction, size depending of number of elements and their set of properties. If you comment this line : 

`graph.command(new OCommandSQL("DELETE VERTEX FROM " + MY_CLASS + " WHERE true")).execute();`

You can see that the time needed to check will not remain unchanged and have the same growth rate, this is what makes me feel that the problem lies within the transaction.

Thank you in advance, Best Regards

Cyprien.

**Comment by CyprienGottstein** [comment]
Hello, any news on that ?

**Comment by tglman** [comment]
Hi @CyprienGottstein,

Yes you are right, the performance depends on the size of the tx, this is because in 3.0.x we made sure that index operations are aware of the changes in transaction, so we send the tx everytime to the server side, we actually do not send it again if the tx didn't change but in your case the tx is actually changing so we need to send it each loop.

We are evaluating how to remove this increased cost, but probably we won't have a solution for the really short term.

Regards

**Comment by CyprienGottstein** [comment]
Hi @tglman ,

I'm not quite catching your explanation " ** we made sure that index operations are aware of the changes in transaction ** ", wouldn't that be way easier to have a local small index into the current transaction and do a local check followed by a global check asking directly the OrientDB server ? I feel like its strange to send the whole payload to perform checks, i haven't committed anything yet...

But, from my understanding of what you are saying, i could do the same thing by creating everything first and once i've created all my objects then i perform the checks on the index and the transaction will be sent only once to the server. Not optimal but it could do the trick.

Anyway, for the time being I simply split into smaller blocs my bulk insert. It's not absolutely critical to me, so i don't mind to wait a bit, just keep me udpated :)

Regards
